### DISCUSSION TITLE
Output comparison vector values and comparison descriptions to reference file

### DISCUSSION THREAD

**Discussion by ianiredan** [moj-analytical-services/splink#1841] [Q&A] [answered]
Hello Splink team,

I am trying to find a way to create a reference sheet that shows the gamma comparison levels and the associated comparison description, or "label_for_charts" entered in the comparison settings for a linkage model.

My use case:
After generating linking predictions, I would QA the predictions in a pandas dataframe, or provide the questionable comparisons for clerical review in a spreadsheet, or a form I throw together to compare the records side-by-side. To review similar comparisons, we would usually review records from one comparison vector together so we are reviewing similar types of matches at once (e.g. gamma_FullName = 19, gamma_DateOfBirth = 4, gamma_Address = 4). 

While the linker predictions show the gamma values, these are numbered values, and we do not remember or the clerical team would not know what each of these numbers mean. To remind myself, I would refer to the linker.match_weights_chart(), or the splink comparison viewer to see this is the "Reversed Month Day", however the model parameter chart does not always show the gamma number in the chart.

What I would like to is to output a list containing gamma level and description, for easy reference, and so I can add these descriptions into the outputs provided for clerical review. Is there a way I can achieve this?

I hope that what I am looking for makes sense, and apologies if I have misused any terminology, or missed how to do this in the documentation. 

Thank you for your work on this product,

Ian

**Comment by RobinL** [comment]
There's no automatic method to do this, but you can do it with a little manual data munging:

In your settings dictionary, make sure you set:
```
    "retain_matching_columns": True,
    "retain_intermediate_calculation_columns": True,
```

and then you can get a lookup of the correspondence between the `gamma_` columns in `predict().as_pandas_dataframe()` and the labels from:

```
linker._settings_obj._as_completed_dict()
```

**Reply by ianiredan** [reply, marked as answer]
Hi Robin,

Thanks for the hint. I'm not a python expert, so with a bit of help from ChatGPT, I was able to work it out. 

In case it is helpful for others, I have copied the code below:

For a basic translation table with `gamma_<column_name>` , `label_for_charts` and `comparison_vector_value`s:
```
# Save the linker settings dictionary
settings_dict = linker._settings_obj._as_completed_dict()

# Extract information and create a mapping DataFrame
mapping_data = []
for comparison in settings_dict['comparisons']:
    column_name = comparison['column_name']
    for level in comparison['comparison_levels']:
        comparison_vector_value = level['comparison_vector_value']
        label_for_charts = level['label_for_charts']
        mapping_data.append({
            'gamma_column': f'gamma_{column_name}',
            'label_for_charts': label_for_charts,
            'comparison_vector_value': comparison_vector_value,
        })
mapping_df = pd.DataFrame(mapping_data)
# drop duplicate rows, if any
mapping_df = mapping_df.drop_duplicates()
```


To add columns `gamma_<column name>_label` after creating a pandas dataframe `df_predictions = predictions.as_pandas_dataframe()` :
```
# Create a mapping for each gamma_<column_name>
label_mappings = {}
for comparison in settings_dict['comparisons']:
    column_name = comparison['column_name']
    gamma_column = f'gamma_{column_name}'
    label_mapping = {level['comparison_vector_value']: level['label_for_charts']
                     for level in comparison['comparison_levels']}
    label_mappings[gamma_column] = label_mapping

# Create new columns in df for each gamma_<column_name>
for gamma_column, mapping in label_mappings.items():
    label_column = f'{gamma_column}_label'
    df_predictions[label_column] = df_predictions[gamma_column].map(mapping)
```

**Reply by RobinL** [reply]
Nice, appreciate you taking the time to post the answer
